File: csharp/binary-search/BinarySearch.cs
using System;

public class BinarySearch
{
    private int[] array;

    public BinarySearch(int[] input)
    {
        array = input;
        Array.Sort(array);
    }

    public int Find(int value)
    {
        int start = 0;
        int end = array.Length - 1;
        int index;

        if (array.Length == 0 || value < array[0] || value > array[end])
        {
            return -1;
        }

        while (start <= end)
        {
            index = (start + end) / 2;

            if (array[index] == value)
            {
                return index;
            }
            else if (array[index] > value)
            {
                end = index - 1;
            }
            else
            {
                start = index + 1;
            }
        }

        return -1;
    }
}

File: fsharp/binary-search/BinarySearch.fs
module BinarySearch

let find input value: int option =
    let rec search input value start stop =
        if start > stop then 
            None 
        else
            let middle = (start + stop) / 2
            match input |> Array.item middle with
            | el when value > el -> search input value (middle + 1) stop
            | el when value < el -> search input value start (middle - 1)
            | _ -> Some middle

    search input value 0 (Array.length input - 1)